src/lib/forwardRefFactory.js

```diff
@@ -0,0 +1,47 @@
+import hoistStatics from 'hoist-non-react-statics'
+import _ from 'lodash'
+import React, { forwardRef } from 'react'
+
+import Ref from 'src/addons/Ref/Ref'
+import { isClassComponent } from './componentUtils'
+
+/**
+ * Use just a string for now (react 16.3), since react doesn't support Symbols in props yet
+ * https://github.com/facebook/react/issues/7552
+ * @type {String}
+ */
+export const forwardRefSymbol = '__forwardRef__'
+
+/**
+ * Creates a function that will choose how to pass a ref.
+ *
+ * @param {Function|Component} Component A Component to wrap
+ * @return {Function}
+ */
+export const forwardFunctionFactory = Component => (props, ref) => {
+  // eslint-disable-next-line react/prop-types
+  if (_.isUndefined(props.as) || _.isString(props.as) || isClassComponent(props.as)) {
+    return <Component {...{ [forwardRefSymbol]: ref, ...props }} />
+  }
```

Comment:
What do you think of explicitly handling the ref for supported use cases and returning the Component in all other cases?:

```jsx
import { isStatelessComponent, supportsRef } from './componentUtils.js'

export const forwardFunctionFactory = Component => (props, ref) => {
  /* eslint-disable react/prop-types */

  if (isStatelessComponent(props.as)) {
    return (
      <Ref innerRef={ref}>
        <Component {...props} />
      </Ref>
    )
  }

  if (supportsRef(props.as)) {
    return <Component {...{ [forwardRefSymbol]: ref, ...props }} />
  }

  return Component
}
```